Refactor ComputeTimeDerivative

nilsdeppe opened the pull request:
## Proposed changes

**I hope using the "Hide Whitespace Changes" on this PR will be helpful**

This does several refactors, I've kept the commits of "This moves code" and "This refactors code" separately to make the review easier, so a lot of smaller commits. The overall goals and motivations here are:
- Make `ComputeTimeDerivative.hpp` shorter (it was getting to ~1500 lines)
- Avoid passing the DataBox to functions that do a lot of work, e.g. calling into the evolution system specific functions. The goal is to be able to have explicit instantiations of these functions. I've done this for the factor `volume_terms` function for some systems, but others have extra template parameters that I think should be cleaned up first.
- Eliminate interface tags from `ComputeTimeDerivative`. They cause inconsistencies with the metric identity satisfying Jacobians and also are very expensive to compile. I'll remove them from `ApplyBoundaryCorrections` in a separate short PR
- I fix one typo-ish issue in `Mesh.cpp` where something is computed outside an `ASSERT` but is only used in the `ASSERT`.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Factor functions out of ComputeTimeDerivative

Moves some functions and metafunctions into separate files to keep the
ComputeTimeDerivative file from growing even more.

nilsdeppe committed:
Factor volume_terms function from ComputeTimeDerivative

Move the function to an impl file to separate the code out a bit.

nilsdeppe committed:
Factor internal mortar data computation from ComputeTimeDerivative

Moves the function into an impl file to keep ComputeTimeDerivative.hpp from
growing too large.

nilsdeppe committed:
Add normal covector quantities to external boundaries

This makes two changes that are closely related:
- adds the normal covector quantities to the external boundaries (done in
  Initialization/Mortars.cpp). This is necessary because the normal covector
  quantities are also needed when applying boundary conditions on external
  boundaries.
- moves the corresponding tags from evolution::dg::Tags::InternalFaces to
  evolution::dg::Tags since they are no longer only used on the internal faces.

nilsdeppe committed:
Stop adding mortars for external boundaries

We won't need them since all boundary conditions will be applied before lifting
neighbor data.

nilsdeppe committed:
Remove most db::get calls from volume_terms computation

Having the volume_terms function not depend on the DataBox will mean we can do
explicit instantiations.

nilsdeppe committed:
Make volume terms computation DataBox-independent

Having the volume_terms function not depend on the DataBox will mean we can do
explicit instantiations.

nilsdeppe committed:
Only sort slice_through dims in Mesh in debug builds

nilsdeppe committed:
Slice mesh directly in ComputeTimeDerivative

nilsdeppe committed:
Reduce DataBox usage in internal_mortar_data

nilsdeppe committed:
Add function to project a single tensor

nilsdeppe committed:
Project mesh velocity to interfaces

nilsdeppe committed:
Project unnormalized normal vectors to interfaces

For trivial maps this might be a bit more expensive than actually evaluating the
map, but is necessary for satisfying the metric identities and allows us to
remove interface (compute) tags.

nilsdeppe committed:
Pass primitive variables to internal_mortar_data

nilsdeppe committed:
Make internal_mortar_data_impl DataBox-independent

I find this makes it easier to understand what the code modifies, and will allow
us to do explicit instantiations for different systems and boundary corrections.

nilsdeppe committed:
Add explicit instantiations of volume terms impl function

- Factor the VolumeTermsImpl.hpp into a tpp so we can do explicit
  instantiations
- Add explicit instantiations for some systems. The hydro systems are templated
  on the equation of state type, which makes explicit instantiations a bit
  tedious (though worthwhile in the long run). However, it looks like the
  template parameter can be replaced with just the thermodynamic dim, which
  would simplify things a lot.

nilsvu commented on src/NumericalAlgorithms/Spectral/Mesh.cpp:
```diff
@@ -34,11 +34,14 @@ template <size_t SliceDim, Requires<(SliceDim <= Dim)>>
+        std::sort(sorted_dims.begin(), sorted_dims.end());
+        auto last_unique = std::unique(sorted_dims.begin(), sorted_dims.end());
+        return last_unique == sorted_dims.end();
+      }(),
```
If I understand correctly, the commit message doesn't fit with what this commit does. Should it perhaps read "Only sort slice_through dims in Debug builds"?

kidder replied:
I think it just suppresses unused variable warning not in debug builds...  But I agree the commit message should be improved

nilsvu reviewed (commented):
not a review, just noticed this

kidder commented on src/Evolution/DiscontinuousGalerkin/Actions/NormalCovectorAndMagnitude.hpp:
```diff
@@ -16,15 +16,13 @@
 #include "Evolution/DiscontinuousGalerkin/NormalVectorTags.hpp"
 #include "Utilities/Gsl.hpp"
 #include "Utilities/TMPL.hpp"
 #include "Utilities/TypeTraits/CreateHasTypeAlias.hpp"
```
remove

kidder commented on src/Evolution/DiscontinuousGalerkin/Actions/InternalMortarDataImpl.hpp:
```diff
@@ -173,8 +173,7 @@ void internal_mortar_data(
       // optimization. This will require making the
```
typo in commit message computed -> compute

nilsdeppe replied:
In the future, could you copy-paste part of the beginning of the commit message? This comment shows up in a few different commits

kidder commented on src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp:
```diff
@@ -202,15 +196,19 @@ void ApplyBoundaryCorrections<Metavariables>::apply_global_time_stepping(
+                << direction
+                << " in the ApplyBoundaryCorrections action. Boundary "
+                   "conditions are applied in the ComputeTimeDerivative action "
+                   "instead. You may have ");
```
You left me in suspense of what I have...

fmahebert commented on src/Evolution/DiscontinuousGalerkin/Actions/ComputeTimeDerivative.hpp:
```diff
@@ -375,19 +378,32 @@ ComputeTimeDerivative<Metavariables>::apply(
+           db::get<::domain::Tags::InverseJacobian<volume_dim, Frame::Logical,
+                                                   Frame::Inertial>>(box),
+       &mesh,
+       &mesh_velocity = db::get<::domain::Tags::MeshVelocity<volume_dim>>(box),
```
Why are you grabbing all these Databox items here, then capturing in the lambda, then using them as non-Databox args in the `mutate_apply` call? Is this not exactly what `mutate_apply`'s `ArgumentTags` are for? Could you defined a new list similar to
```
using every_arg_tags = tmpl::list<
    ::domain::Tags::DivMeshVelocity,
    variables_tag,
    domain::Tags::Coordinates<volume_dim, Frame::Inertial>,
    ::domain::Tags::InverseJacobian<volume_dim, Frame::Logical, Frame::Inertial>,
    ::domain::Tags::MeshVelocity<volume_dim>,
    compute_volume_time_derivative_terms::argument_tags
    >;
```
to use on line 383, then avoid capturing all these in the lambda, then change the argument order in the `volume_terms` function?

nilsdeppe replied:
Because then I have to deal with that `mutate_apply` forces an argument ordering on me, and that doesn't work with the pointer parts, or I have to reorder all the arguments so that maybe it works. I also can't split a parameter pack, which makes that switch impossible in this case.

fmahebert replied:
Okay, fair enough

fmahebert commented on src/Evolution/DiscontinuousGalerkin/Actions/VolumeTermsImpl.hpp:
```diff
@@ -114,82 +112,65 @@ void volume_terms(
+  // compute_volume_time_derivative_terms::apply call.
+  (void)partial_derivs;
+  (void)volume_fluxes;
+  (void)temporaries;
```
I'd favor a consistent use of either `[[maybe_unused]]` or `(void)blah` to address this problem ... or are these accomplishing different tasks?

nilsdeppe replied:
I think these can be updated now, but I haven't figured out how to use `[[maybe_unused]]` with lambda captures, which sometimes forces this non-uniform behavior

Edit: I also generally prefer scoping the `[[maybe_unused]]`, at least in production, which is why I use the ` else { (void)blah; }` approach.

fmahebert replied:
Got it, thanks for clarifying. Since the scopes differ I'd be okay leaving as-is in spite of my preference for uniform styling :P

fmahebert commented on src/NumericalAlgorithms/Spectral/Mesh.cpp:
```diff
@@ -34,11 +34,14 @@ template <size_t SliceDim, Requires<(SliceDim <= Dim)>>
+        auto last_unique = std::unique(sorted_dims.begin(), sorted_dims.end());
+        return last_unique == sorted_dims.end();
+      }(),
+      "Dimensions to slice through contain duplicates.");
```
Not asking you to change this, but next time this commit "Only sort slice_through dims in Mesh" (as others have said, needs rephrasing) would be better presented in a small bugfix PR

fmahebert commented on src/Evolution/DiscontinuousGalerkin/NormalVectorTags.hpp:
```diff
@@ -13,8 +13,8 @@
 
-/// %Tags used on the interior faces of the elements
-namespace evolution::dg::Tags::InternalFace {
+/// %Tags used on the faces of the elements
```
The namespace `evolution::dg::Tags` already has the Doxygen description "%Tags used for DG evolution scheme." in file `Evolution/DiscontinuousGalerkin/MortarTags.hpp`.

This existing documentation is much more appropriate than what you're adding here, because the namespace has a broad name that implies no restriction to tags/faces.

I think the todo items here are,
- delete this Dox line here in this commit
- make a todo issue (or add a commit in this PR) so that the Doxygen documentation of the `evolution::dg::Tags` namespace is in a more appropriate file than the current `MortarTags.hpp`, perhaps in a new `Evolution/DiscontinuousGalerkin/Tags.hpp`

fmahebert commented on src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.cpp:
```diff
@@ -70,6 +67,10 @@ auto Mortars<Dim, AddFluxBoundaryConditionMortars>::apply_impl(
+  for (const auto& direction : element.external_boundaries()) {
+    normal_covector_quantities[direction] = std::nullopt;
+  }
+
```
Please separate the functionality change and the Tag namespace changes into different commits

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/Initialization/Test_Mortars.cpp:
```diff
@@ -188,10 +187,11 @@ void test<1>(const Spectral::Quadrature quadrature) {
+             Variables<tmpl::list<evolution::dg::Tags::MagnitudeOfNormal,
+                                  evolution::dg::Tags::NormalCovector<1>>>>>
+      expected_normal_covector_quantities{{Direction<1>::lower_xi(), {}},
+                                          {Direction<1>::upper_xi(), {}}};
```
Please separate the functionality change and the Tag namespace changes into different commits

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/Initialization/Test_Mortars.cpp:
```diff
@@ -250,11 +250,13 @@ void test<2>(const Spectral::Quadrature quadrature) {
+      expected_normal_covector_quantities{{Direction<2>::lower_xi(), {}},
+                                          {Direction<2>::upper_xi(), {}},
+                                          {Direction<2>::lower_eta(), {}},
+                                          {Direction<2>::upper_eta(), {}}};
```
again

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/Initialization/Test_Mortars.cpp:
```diff
@@ -322,12 +324,13 @@ void test<3>(const Spectral::Quadrature quadrature) {
+      expected_normal_covector_quantities{
+          {Direction<3>::lower_xi(), {}},   {Direction<3>::upper_xi(), {}},
+          {Direction<3>::lower_eta(), {}},  {Direction<3>::upper_eta(), {}},
+          {Direction<3>::lower_zeta(), {}}, {Direction<3>::upper_zeta(), {}}};
```
again

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -244,4 +244,56 @@ void project_tensors_to_boundary(
 }
+
+/*!
+ * \brief Projects a tensor to the face
```
Not asking you to change this, but this commit "Add function to project a single tensor" would have been better in a separate small PR

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -244,4 +244,56 @@ void project_tensors_to_boundary(
+  const Mesh<Dim> uniform_gauss_mesh(volume_mesh.extents(0),
+                                     volume_mesh.basis(0),
+                                     Spectral::Quadrature::Gauss);
+  if (volume_mesh.quadrature() == uniform_gauss_mesh.quadrature()) {
```
You're creating this `uniform_gauss_mesh` but then are only checking the quadrature part of it.

Either,
1) if you only want to check the quadrature, then write
```
if (volume_mesh.quadrature() == make_array<Dim>(Spectral::Quadrature::Gauss)) {
```
and delete the `uniform_gauss_mesh` variable... OR
2) if you want to check you have a uniform gauss mesh, then write
```
if (volume_mesh == uniform_gauss_mesh) {
```

Furthermore, this `if` is checking only that you have Gauss points, but the ASSERT in the `else` branch is much stricter: it checks you have a uniform extents, uniform basis, and uniform GL quadrature. Is it intentional that the `else` branch is so much stricter?

fmahebert commented on src/Evolution/DiscontinuousGalerkin/Actions/InternalMortarDataImpl.hpp:
```diff
@@ -186,12 +180,55 @@ void internal_mortar_data(
-                make_not_null(&fields_on_face), local_direction,
-                unnormalized_normal_covectors, moving_mesh_map);
+            const bool mesh_is_moving = not moving_mesh_map.is_identity();
+            if (auto& normal_covector_quantity =
```
Can this be const?

nilsdeppe replied:
no, it's assigned to on line 191:
```cpp
              if (not normal_covector_quantity.has_value()) {
                normal_covector_quantity = Variables<
                    tmpl::list<evolution::dg::Tags::MagnitudeOfNormal,
                               evolution::dg::Tags::NormalCovector<Dim>>>{
                    fields_on_face.number_of_grid_points()};
              }
```
Note that this is actually the thing that the `db::mutate` call is mutating :)

fmahebert commented on src/Evolution/DiscontinuousGalerkin/Actions/InternalMortarDataImpl.hpp:
```diff
@@ -186,12 +180,55 @@ void internal_mortar_data(
+                for (auto& normal_covector_component :
+                     get<evolution::dg::Tags::NormalCovector<Dim>>(
+                         *normal_covector_quantity)) {
+                  normal_covector_component *= sign;
```
I find the logic is easier to follow written as,
```
if (local_direction.side() == Side::Lower) {
  for (...) {
    normal_covector_component *= -1.0;
  }
}
```

What do you think?

nilsdeppe replied:
I intentionally didn't do it that way so as to not assume what the implementation of `sign()` does. If both you and @kidder would like it changed I'll change it, or I can add a comment to clarify why it's done this way

fmahebert commented on src/Evolution/DiscontinuousGalerkin/Actions/PackageDataImpl.hpp:
```diff
@@ -53,13 +53,34 @@ double dg_package_data(
+        mesh_velocity,
+    const db::DataBox<DbTagsList>& box, tmpl::list<VolumeTags...> /*meta*/,
+    tmpl::list<ProjectedFieldTagsForCorrection...> /*meta*/) noexcept {
+  return dg_package_data<System>(
```
In the `internal_mortar_data` case, you renamed the non-Databox version with an `_impl` suffix. Should the same be done here? Or am I missing something about the interfaces (e.g., is the Databox interface for `dg_package_data` still called directly?)?

nilsdeppe replied:
The DataBox interface is still called in the boundary conditions at this point. I haven't finalized that code yet. I'm hoping to be able to get rid of the `DataBox` one completely, if possible. Since they both currently act as a semi-public interface, or at least are on equal footing in terms of whether they should be called directly, I decided to not add `_impl`. Does that sound okay?

fmahebert commented on src/Evolution/DiscontinuousGalerkin/Actions/ComputeTimeDerivative.hpp:
```diff
@@ -342,12 +342,15 @@ ComputeTimeDerivative<Metavariables>::apply(
       typename system::compute_volume_time_derivative_terms;
 
   const Mesh<volume_dim>& mesh = db::get<::domain::Tags::Mesh<volume_dim>>(box);
+  const ::dg::Formulation& dg_formulation =
```
- should not need to make a reference for an enum
- check for other places below that grab the enum from the Databox, that should now just use your new variable (e.g. line 375)

fmahebert commented on src/Evolution/DiscontinuousGalerkin/Actions/InternalMortarDataImpl.hpp:
```diff
@@ -173,8 +173,7 @@ void internal_mortar_data(
       // compute tag.
-      db::mutate<
-          evolution::dg::Tags::InternalFace::NormalCovectorAndMagnitude<Dim>>(
+      db::mutate<evolution::dg::Tags::NormalCovectorAndMagnitude<Dim>>(
```
This commit "Add normal covector quantities to external boundaries" appears to be 95% changing the namespace for some Tags, and 5% behavior changes. Please split it into two commits.

nilsdeppe replied:
Without the functionality changes the previous namespace was "more correct" It's the functionality changes that is triggering the rename, that's why they are in the same commit (and why I think they belong in the same commit). Thoughts?

fmahebert replied:
I'd be okay leaving the commit as-is if you improve the documentation of it by changing the commit message so (in the lengthy part) it has two bullets... one bullet explaining the functionality change and one bullet explaining the namespace change. How does that sound?

nilsdeppe committed:
Make volume_terms depend on ComputeTimeDerivativeTerms

This makes explicit instantiations in some cases somewhat easier, and it
minimizes how much volume_terms depends on the System (almost not at all).

nilsdeppe committed:
Stop requiring uniform mesh for projection

We only actually care if the mesh in the direction is Gauss or Gauss-Lobatto,
the perpendicular directions are irrelevant

nilsdeppe commented:
I have done the following:
- rebased on develop
- added fixup commits (I added these inline, dealing with conflicts in later commits immediately)
- reworded commit messages
- added the commit [Stop requiring uniform mesh for projection](https://github.com/sxs-collaboration/spectre/pull/2799/commits/ecce7596efb9a72e3a5877b646a0d2b43574ca86)
- added the commit [Make volume_terms depend on ComputeTimeDerivativeTerms](https://github.com/sxs-collaboration/spectre/pull/2799/commits/669ef44e420df8c7e032855933eab51dbb78c9b5)

Note that the links will be invalid once I rebase again :(

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -273,12 +271,6 @@ void project_tensor_to_boundary(
-           "Gauss or Gauss-Lobatto mesh, but got "
-               << volume_mesh);
-
     const size_t sliced_dim = direction.dimension();
```
Nitpick: move this outside the `if/else` block to avoid calling `direction.dimension()` several times in the `if` branch

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -111,11 +109,6 @@ void project_contiguous_data_to_boundary(
-           "The current implementation assumes the mesh be either a uniform "
-           "Gauss or Gauss-Lobatto mesh, but got "
-               << volume_mesh << ".");
     const size_t sliced_dim = direction.dimension();
```
same nitpick

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -203,12 +194,6 @@ void project_tensors_to_boundary(
-           "Gauss or Gauss-Lobatto mesh, but got "
-               << volume_mesh);
-
     const size_t sliced_dim = direction.dimension();
```
same nitpick

fmahebert commented on src/Evolution/Executables/NewtonianEuler/VolumeTermsInstantiation.cpp:
```diff
@@ -0,0 +1,92 @@
+                        (::NewtonianEuler::Solutions::LaneEmdenStar), (0))
+
+#undef INSTANTIATION
+#undef INITIAL_DATA_TYPE
```
`#undef SYSTEM` as well?

fmahebert commented on src/Evolution/Executables/NewtonianEuler/CMakeLists.txt:
```diff
@@ -1,6 +1,35 @@
+# depend on the system for the tags. This cyclic dependency is already present
+# in the code, but currently only through header files. The same cyclic
+# dependency exists between NewtonianEulerSources and NewtonianEuler. To resolve
+# these, we need to factor the NewtonianEuler tags into a separate library.
```
I suggest "we need to" -> "we will need to" to make clear you're describing a future TODO (and not describing why the code is the way it is today).

Perhaps also add a suggestion that the instantiations be moved back into the System directory (like for other systems) after the circular dep is broken? Perhaps "After the NewtonianEuler tags are factored in this way, the instantiation source file can be moved to the NewtonianEuler library" or similar?

fmahebert commented on src/Evolution/Executables/NewtonianEuler/VolumeTermsInstantiation.cpp:
```diff
@@ -0,0 +1,92 @@
+// used in code that will be removed and that we don't care about. Even the
+// dependence on the thermodynamic dimension can probably be removed. In either
+// case, we don't care about the thermodynamic dimension for the explicit
+// instantiation either.
```
This comment left me with so many questions...
- "The system is currently templated..." ... which system? 
- "...used in code that will be removed..." ... which code?
- "Even the dependence on the thermodynamic dimension" ... the dependence of what code on the thermodynamic dim?

I'm not sure the comment needs changing. If you reply in Github with a little more detail and explain for me what the "code that will be removed and that we don't care about" is, I'll probably be happy as-is. But you could also expand a bit in code comment if you think that'd be generally more helpful (depends how soon the other code gets deleted probably?)

nilsdeppe replied:
The evolution systems are sometimes templated on the EOS because the char speed compute tag needs them. This makes explicit instantiations need some form of EOS, even though the parts the volume_terms instantiation cares about doesn't depend on the EOS. It probably will depend on the thermodynamic dim, i.e. does the pressure depend on just density or density and internal energy. So to answer your questions:
- evolution system
- compute_char_speed
- the evolution system might not even need to depend on the thermodynamic dim. This isn't clear yet, but I think we can template the `compute_time_derivative_terms::apply` on the thermodynamic dim instead. We probably need to decide how to do prim recovery to answer this question for GRMHD. As in, do we want runtime choosing of the recovery scheme and the EOS

fmahebert replied:
Okay, thanks for clarifying. I'd be okay leaving the comment as-is.

fmahebert commented on src/Evolution/Systems/GrMhd/ValenciaDivClean/VolumeTermsInstantiation.cpp:
```diff
@@ -0,0 +1,74 @@
+// used in code that will be removed and that we don't care about. Even the
+// dependence on the thermodynamic dimension can probably be removed. In either
+// case, we don't care about the thermodynamic dimension for the explicit
+// instantiation either.
```
same comment

fmahebert commented on src/Evolution/Systems/RelativisticEuler/Valencia/VolumeTermsInstantiation.cpp:
```diff
@@ -0,0 +1,76 @@
+// used in code that will be removed and that we don't care about. Even the
+// dependence on the thermodynamic dimension can probably be removed. In either
+// case, we don't care about the thermodynamic dimension for the explicit
+// instantiation either.
```
again

fmahebert reviewed (commented):
I am happy with commit rewordings and fixups (and your comments explaining why you're not implementing some of my requests). You have my go-ahead to squash.

Here are a few nitpicks or clarifications in the new commits, you can squash any changes in directly while squashing the other things.

Thanks!

kidder reviewed (commented):
I am fine with squashing

nilsdeppe commented:
I've rebased and squashed in the changes. There was a missing `undef SYSTEM` in Rel Euler that I also added.

Thanks for the reviews :)
